ActiveLabel/RegexFactory.swift

```diff
@@ -0,0 +1,70 @@
+        "((https?://|www\\.|pic\\.)[-\\w;/?:@&=+$\\|\\_.!~*\\|'()\\[\\]%#,☺]+[\\w/#](\\(\\))?)" +
+    "(?=$|[\\s',\\|\\(\\).:;?\\-\\[\\]>\\)])"
+    private static var cachedRegularExpressions: [ActiveType : NSRegularExpression] = [:]
+
+    func getElements(from text: String, with pattern: ActiveType, range: NSRange) -> [NSTextCheckingResult]{
+        guard let regexForElement = getRegex(forActiveType: pattern) else { return [] }
+        return regexForElement.matches(in: text, options: [], range: range)
+    }
+
+    private func getRegex(forActiveType activeType: ActiveType) -> NSRegularExpression? {
+        // TODO: UT the equality of two custom patterns
+        if let cachedRegex = RegexFactory.cachedRegularExpressions[activeType] {
+            return cachedRegex
+        }
+        
+        switch activeType {
+        case .phone:
+            let types: NSTextCheckingResult.CheckingType = [.phoneNumber]
+            guard let createdRegex = try? NSDataDetector(types: types.rawValue) else { return nil }
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .address:
+            let types: NSTextCheckingResult.CheckingType = [.address]
+            guard let createdRegex = try? NSDataDetector(types: types.rawValue) else { return nil }
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .date:
+            let types: NSTextCheckingResult.CheckingType = [.date]
+            guard let createdRegex = try? NSDataDetector(types: types.rawValue) else { return nil }
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .url:
+            let createdRegex = try? NSRegularExpression(pattern: urlPattern, options: [.caseInsensitive])
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .email:
+            let createdRegex = try? NSRegularExpression(pattern: emailPattern, options: [.caseInsensitive])
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .mention:
+            let createdRegex = try? NSRegularExpression(pattern: mentionPattern, options: [.caseInsensitive])
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .hashtag:
+            let createdRegex = try? NSRegularExpression(pattern: hashtagPattern, options: [.caseInsensitive])
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
+        case .custom(pattern: let pattern):
+            let createdRegex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
+            RegexFactory.cachedRegularExpressions[activeType] = createdRegex
+            return createdRegex
```

Comment:
we can refactor out these 2 repeating lines of code

Reply:
what benefits a class instance of RegexParser provides over a static instance?

Reply:
>we can refactor out these 2 repeating lines of code

👍

>what benefits a class instance of RegexParser provides over a static instance?

It's finally injectable and testable